File: hermes_ng/schema/actions.py
from pydantic import BaseModel, Field
from typing import Optional


class BaseAction(BaseModel):
    action_id: str = Field(..., description="唯一动作标识符")
    reasoning: str = Field(..., description="执行此动作的思考路径")
    tool_name: str = Field(..., description="调用的工具名称")

    class Config:
        extra = "forbid"


class MemoryWriteAction(BaseAction):
    tool_name: str = "memory_write"
    key: str = Field(..., description="记忆键名")
    value: str = Field(..., description="记忆内容")


class MemoryReadAction(BaseAction):
    tool_name: str = "memory_read"
    key: str = Field(..., description="要读取的记忆键名")


class SkillCreateAction(BaseAction):
    tool_name: str = "skill_create"
    name: str = Field(..., description="技能名称")
    code: str = Field(..., description="技能代码内容")
    boundary: str = Field(default="", description="边界定位信息")


class SkillUpdateAction(BaseAction):
    tool_name: str = "skill_update"
    name: str = Field(..., description="技能名称")
    code: str = Field(..., description="更新后的技能代码")


class BashRunAction(BaseAction):
    tool_name: str = "bash_run"
    command: str = Field(..., description="要执行的bash命令")
    workdir: Optional[str] = Field(None, description="工作目录")


class FileReadAction(BaseAction):
    tool_name: str = "file_read"
    path: str = Field(..., description="文件路径")


class FileWriteAction(BaseAction):
    tool_name: str = "file_write"
    path: str = Field(..., description="文件路径")
    content: str = Field(..., description="文件内容")


class FinishAction(BaseAction):
    tool_name: str = "finish"
    summary: str = Field(..., description="任务完成总结")

File: zilli-rs/src/schema/actions.rs
use serde::{Deserialize, Serialize};
use std::collections::HashMap;

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize, strum::Display)]
#[serde(rename_all = "snake_case")]
#[strum(serialize_all = "snake_case")]
pub enum ActionType {
    MemoryWrite,
    MemoryRead,
    SkillCreate,
    SkillUpdate,
    BashRun,
    FileRead,
    FileWrite,
    Finish,
    Think,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct BaseAction {
    pub action_id: String,
    pub reasoning: Option<String>,
    pub tool_name: ActionType,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct MemoryWriteAction {
    pub action_id: String,
    pub reasoning: Option<String>,
    pub key: String,
    pub value: serde_json::Value,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct MemoryReadAction {
    pub action_id: String,
    pub reasoning: Option<String>,
    pub key: String,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct SkillCreateAction {
    pub action_id: String,
    pub reasoning: Option<String>,
    pub name: String,
    pub code: String,
    pub boundary: Option<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct SkillUpdateAction {
    pub action_id: String,
    pub reasoning: Option<String>,
    pub name: String,
    pub code: String,
    pub boundary: Option<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct BashRunAction {
    pub action_id: String,
    pub reasoning: Option<String>,
    pub command: String,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct FileReadAction {
    pub action_id: String,
    pub reasoning: Option<String>,
    pub path: String,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct FileWriteAction {
    pub action_id: String,
    pub reasoning: Option<String>,
    pub path: String,
    pub content: String,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct FinishAction {
    pub action_id: String,
    pub reasoning: Option<String>,
    pub summary: Option<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct TaskConfig {
    pub id: String,
    pub name: String,
    pub description: String,
    pub category: String,
    pub max_steps: i32,
    pub initial_context: Option<String>,
    pub verification: Option<serde_json::Value>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct RewardRule {
    #[serde(rename = "type")]
    pub rule_type: String,
    pub weight: f64,
    pub params: Option<HashMap<String, serde_json::Value>>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct TrajectoryTemplateStep {
    pub step_number: i32,
    pub action_type: String,
    pub required_fields: Vec<String>,
    pub expected_observation: Option<String>,
}
